xarray/core/dataarray.py

```diff
@@ -462,8 +460,8 @@ def __init__(
 
         # These fully describe a DataArray
         self._variable = variable
-        assert isinstance(coords, dict)
-        self._coords = coords
+        # Variables can have undefined data type according to MyPy due to the fastpath
+        self._coords = cast(dict, coords)
```

Comment:
I think mypy is correct to complain here. Set `fastpath=True` it becomes strange to not have the assert. 

Remember that `coords: Sequence[Sequence | pd.Index | DataArray] | Mapping | None = None` at this point.

Reply:
Ok i added back the asserts. they did not have a significant effect on the benchmark I was running.